### PR TITLE
Turn on assertion reachability checks by default

### LINKED ISSUE #637: RMC should distinguish between reachable assertions that hold and unreachable ones
Requested feature: Consider the following example:
```rust
fn foo(x: u8) -> u8 {
    if x < 10 {
        17
    } else {
        5
    }
}

#[no_mangle]
fn main() {
    let x: u8 = rmc::nondet();
    let y = foo(x);
    if y > 200 {
        assert!(x != y);
    } else if y > 100 {
        assert!(x == y);
    } else {
        assert!(y != 10);
    }
}
```
RMC reports the following when run with `rmc test.rs`:
```
/home/ubuntu/tools/rmc/src/test/report/vacuity/test.rs function main
[main.assertion.1] line 14 assertion failed: x != y: SUCCESS
[main.assertion.2] line 16 assertion failed: x == y: SUCCESS
[main.assertion.3] line 18 assertion failed: y != 10: SUCCESS
...
** 0 of 66 failed (1 iterations)
VERIFICATION SUCCESSFUL
```
This output is misleading in the following sense:
1. Even though the first assert holds, it is actually unreachable
2. The second assert doesn't hold and is unreachable, but there's no indication of its unreachability in the output

The third assert is the only reachable one, so it's the only one that it makes sense to report "SUCCESS" for.

It would be useful if RMC automatically generates reachability checks for such asserts and takes their result into account when reporting results for asserts: if an assert is unreachabe, it would be reported as UNREACHABLE and not SUCCESS/FAILURE.

This can perhaps be put under an option that can be turned on/off, similar to CBMC's `--cover assertions` option (so that the user can turn them off if performance is an issue).

### PR DESCRIPTION
**Pull request by zhassan-aws**
### Description of changes: 

Turn on assertion reachability checks by default, and replace the option `--assertion-reach-checks` with `--no-assertion-reach-checks` that can be used for turning them off.

### Resolved issues:

Resolves #637 

### Call-outs:

Since reachability checks add some annotations to each check/assert's description, e.g.
```
[KANI_CHECK_ID_a.c64ace8a::a_0] attempt to add with overflow: FAILURE
```
the CBMC result parser **must** now be run in order to clean those up. This makes the old format pretty much unusable at this point since it generates output that is not user consumable. Perhaps we should rename it from "old" to "raw" or something similar.

### Testing:

* How is this change tested? Regressions have been updated

* Is this a refactor change? No

### Checklist
- [X] Each commit message has a non-empty body, explaining why the change was made
- [X] Methods or procedures are documented
- [X] Regression or unit tests are included, or existing tests cover the modified code
- [X] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: tests/expected/reach/assert/reachable_fail/test.rs [hunk 1]
```diff
@@ -1,6 +1,6 @@
 // Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 // SPDX-License-Identifier: Apache-2.0 OR MIT
-// kani-flags: --assertion-reach-checks --output-format regular --no-default-checks
+// kani-flags: --output-format regular --no-default-checks
```
Comment (tedinski) [inline tests/expected/reach/assert/reachable_fail/test.rs:3]: do we still need the output-format flag here? (and several more places below...)

(I'm also mildly surprised by the `--no-default-checks` but that seems unrelated: weren't we switching to "regular" because of assertion reachability and now both are default?)

File: tests/kani/DynTrait/drop_nested_boxed_dyn.rs [hunk 2]
```diff
@@ -5,6 +5,12 @@
 // There is an implicit self-recursive call to drop_in_place, so we
 // need to set an unwind bound.
 
+// Temporarily disabling assertion reachability checks because they trigger a
+// crash in CBMC:
+// https://github.com/diffblue/cbmc/issues/6691
+// https://github.com/model-checking/kani/issues/861
+// kani-flags: --no-assertion-reach-checks
```
Comment (tedinski) [inline tests/kani/DynTrait/drop_nested_boxed_dyn.rs:12]: This is a lot of these comment blocks. It looks like by maybe Monday the new github macos image will be deployed (usually it is by now, but apparently it's only at 3% of fleet this week), and that would let us upgrade to 5.52 which supposedly has the fix.

Planning on doing that version bump and fix all these early next week?

File: tests/expected/reach/assert/reachable_fail/test.rs (hunk 1 above)
Comment (zhassan-aws) [reply tests/expected/reach/assert/reachable_fail/test.rs:3]: > do we still need the output-format flag here? (and several more places below...)
> 
No, we no longer need it. I'll remove it.

> (I'm also mildly surprised by the `--no-default-checks` but that seems unrelated: weren't we switching to "regular" because of assertion reachability and now both are default?)
>
The reason this was using `--no-default-checks` was to reduce the number of checks to make the expected tests less likely to pass due to some other check having "SUCCESS". Now that we support multi-line checks, `--no-default-checks` should no longer be needed. I'll remove it.

File: tests/kani/DynTrait/drop_nested_boxed_dyn.rs (hunk 2 above)
Comment (zhassan-aws) [reply tests/kani/DynTrait/drop_nested_boxed_dyn.rs:12]: > Planning on doing that version bump and fix all these early next week?

I have the branch for bumping up the version ready in https://github.com/zhassan-aws/kani/commit/ea2026570beb546afc4b1e447f184da214128a24. I was hoping that 5.52.0 would be available in github today to avoid needing to turn reachability off for those tests. Would you prefer that I hold this PR off until we get the new version of CBMC?

**Review by tedinski — APPROVED** [review]
